### PR TITLE
fix: missing signers' votes

### PR DESCRIPTION
**Pull request by Jiloc**
## Description

Closes: #1237

## Changes
- add new `deposit_decisions_retry_window` param to the signers' settings (defaults to 3)
- add query to get the all the `DepositSigner` entries from the DB from the last N blocks
- add resend mechanism in the request_decider to resend all the Requests of the last 3 blocks when a new bitcoin block arrives
 
## Testing Information
- add query integration tests

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: signer/src/storage/postgres.rs [hunk 1]
```diff
@@ -1921,6 +1921,54 @@ impl super::DbRead for PgStore {
...
+            WHERE deposit_signers.signer_pub_key = $3
+            "#,
+        )
+        .bind(chain_tip)
+        .bind(i32::from(context_window))
+        .bind(signer_public_key)
+        .fetch_all(&self.0)
+        .await
+        .map_err(Error::SqlxQuery)
+    }
```
Comment (Jiloc) [inline signer/src/storage/postgres.rs:1971]: A simple optimization could be to specialize a bit the query and only get decision of the deposits which have less than the total number of signers entries in the table.

File: signer/src/request_decider.rs [hunk 2]
```diff
@@ -248,6 +276,20 @@ where
...
 
+    /// Send the given deposit decisions to the other signers for redundancy.
+    #[tracing::instrument(skip_all)]
+    pub async fn handle_deposit_decisions_to_retry(
+        &mut self,
+        decisions: Vec<model::DepositSigner>,
+        chain_tip: &BitcoinBlockHash,
+    ) -> Result<(), Error> {
+        for decision in decisions.into_iter().map(SignerDepositDecision::from) {
+            self.send_message(decision, chain_tip).await?;
```
Comment (matteojug) [inline signer/src/request_decider.rs:281]: Since this retry is a "best effort" type of thing, should we just log the error in here instead of bailing at the first one?

File: signer/src/storage/postgres.rs (hunk 1 above)
Comment (matteojug) [reply signer/src/storage/postgres.rs:1971]: This is doing something slightly different from what we initially talked about: here we pick the deposits that were confirmed (on the canonical chain) in the last three blocks, but this means that if we fetch from Emily a deposit that was already confirmed but submitted (to emily) only recently we may not retry it.
This is fine wrt the original issue (signers getting a vote for an viewed-as-unconfirmed deposit), but if we were to have a real _retry 3 times_ approach we would also add redundancy to lost votes messages (for other, unrelated reasons) -- but it would require adding more info somewhere else in the db, so maybe it's not really worth it.

Also a possible simplification here could be to ignore the canonical chain, but given the context window is so small the recursive part should be an issue.

File: signer/src/storage/postgres.rs [hunk 3]
```diff
@@ -1921,6 +1921,54 @@ impl super::DbRead for PgStore {
...
+
+                UNION ALL
+
+                -- Recursive member: Fetch the parent block using the last block's parent_hash
+                SELECT parent.block_hash, parent.block_height, parent.parent_hash,
+                       parent.created_at, last.depth + 1
+                FROM sbtc_signer.bitcoin_blocks parent
+                JOIN context_window last ON parent.block_hash = last.parent_hash
+                WHERE last.depth < $2
+            ),
```
Comment (matteojug) [inline signer/src/storage/postgres.rs:1947]: Can we use `bitcoin_blockchain_of` here?

File: signer/src/request_decider.rs [hunk 4]
```diff
@@ -248,6 +276,20 @@ where
...
     }
 
+    /// Send the given deposit decisions to the other signers for redundancy.
+    #[tracing::instrument(skip_all)]
+    pub async fn handle_deposit_decisions_to_retry(
+        &mut self,
+        decisions: Vec<model::DepositSigner>,
+        chain_tip: &BitcoinBlockHash,
+    ) -> Result<(), Error> {
+        for decision in decisions.into_iter().map(SignerDepositDecision::from) {
```
Comment (matteojug) [inline signer/src/request_decider.rs:280]: Should we do this concurrently? Instead of awaiting each one.

File: signer/tests/integration/postgres.rs [hunk 5]
```diff
@@ -1400,6 +1400,71 @@ async fn fetching_deposit_request_votes() {
...
+    let signer_set = testing::wsts::generate_signer_set_public_keys(&mut rng, num_signers);
+
+    let test_data = TestData::generate(&mut rng, &signer_set, &test_model_params);
+    test_data.write_to(&pg_store).await;
+
+    let chain_tip = test_data
+        .bitcoin_blocks
+        .iter()
+        .max_by_key(|x| (x.block_height, x.block_hash))
+        .unwrap();
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1430]: nit: we can use `pg_store.get_bitcoin_canonical_chain_tip`

File: signer/tests/integration/postgres.rs [hunk 6]
```diff
@@ -1400,6 +1400,71 @@ async fn fetching_deposit_request_votes() {
...
+                signer_pub_key: *signer_pub_key,
+                can_accept: true,
+                can_sign: true,
+            };
+            pg_store
+                .write_deposit_signer_decision(&decision)
+                .await
+                .unwrap();
+        }
+    }
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1446]: Note that `TestData::generate` already [generates signers decision](https://github.com/stacks-network/sbtc/blob/main/signer/src/testing/storage/model.rs#L401-L412) (randomly) based on `num_signers_per_request`, and given that `write_deposit_signer_decision` `ON CONFLICT DO NOTHING` this snippet shouldn't have any effect (IIUC)

File: signer/tests/integration/postgres.rs [hunk 7]
```diff
@@ -1400,6 +1400,71 @@ async fn fetching_deposit_request_votes() {
...
+
+    let signer_pub_key = signer_set.first().unwrap();
+
+    let deposit_decisions = pg_store
+        .get_deposit_signer_decisions(&chain_tip.block_hash, 3, signer_pub_key)
+        .await
+        .unwrap();
+
+    assert_eq!(deposit_decisions.len(), 3);
+    for deposit in test_data.deposit_requests[3..].iter() {
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1456]: qq: why the `[3..]`?

File: signer/tests/integration/postgres.rs [hunk 8]
```diff
@@ -1400,6 +1400,71 @@ async fn fetching_deposit_request_votes() {
...
+    }
+
+    let signer_pub_key = signer_set.first().unwrap();
+
+    let deposit_decisions = pg_store
+        .get_deposit_signer_decisions(&chain_tip.block_hash, 3, signer_pub_key)
+        .await
+        .unwrap();
+
+    assert_eq!(deposit_decisions.len(), 3);
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1455]: We should also check that we are fetching only our own (by `signer_pub_key`) decisions.

File: signer/tests/integration/postgres.rs [hunk 9]
```diff
@@ -1400,6 +1400,71 @@ async fn fetching_deposit_request_votes() {
...
+    for deposit in test_data.deposit_requests[3..].iter() {
+        assert!(deposit_decisions
+            .iter()
+            .find(|decision| {
+                decision.txid == deposit.txid && decision.output_index == deposit.output_index
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1460]: This should also check that the matching decision is our own (based on `signer_pub_key`)

File: signer/src/storage/postgres.rs (hunk 1 above)
Comment (Jiloc) [reply signer/src/storage/postgres.rs:1971]: Great thinking on the edge case of missing older transactions! A simple approach that avoids modifying the database could be to use the `created_at` column of the bitcoin block in the `bitcoin_blocks` table relative to the selected context_window. With each new block, we can retrieve the creation time of the block received N-updates ago and fetch all decisions with a `created_at` greater than that block's creation time. The only downside is that if blocks arrive very close to each other, we might end up retrying a vote an additional time. What do you think?

File: signer/src/request_decider.rs (hunk 4 above)
Comment (Jiloc) [reply signer/src/request_decider.rs:278]: Yeah, I thought about it, and got too scared with all the previous interactions with Emily and concurrent requests 🥲 , but in this case it make sense. We are not processing concurrent messages on the P2P network

File: signer/src/storage/postgres.rs (hunk 1 above)
Comment (matteojug) [reply signer/src/storage/postgres.rs:1971]: Oh that's a nice way to avoid extra stuff! And an extra retry shouldn't really be an issue, we would just send 4x more votes message but I don't think that's overwhelming -- but maybe we should double check, just in case.
I'm okay with leveraging `created_at`, but maybe let's check also with @djordon or @cylewitruk , just to be sure it's not an overkill or we are missing some other corner case.

File: signer/tests/integration/postgres.rs [hunk 10]
```diff
@@ -1400,6 +1400,56 @@ async fn fetching_deposit_request_votes() {
...
+    assert_eq!(deposit_decisions.len(), 3);
+    // Test data contains 5 deposit requests, we should get decisions for
+    // the last 3.
+    for deposit in test_data.deposit_requests[2..].iter() {
+        assert!(deposit_decisions.iter().any(|decision| {
+            decision.txid == deposit.txid
+                && decision.output_index == deposit.output_index
+                && decision.signer_pub_key == *signer_pub_key
+        }));
+    }
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1448]: This doesn't always work: `TestData` generates a chain with forks, so picking the last 3 deposits requests doesn't guarantee that those 3 are all on the canonical chain; it does work with `51` as random seed, but if you change that to `52` the test fails.

File: signer/src/testing/storage/model.rs [hunk 11]
```diff
@@ -515,6 +524,8 @@ pub struct Params {
     pub num_withdraw_requests_per_block: usize,
     /// The number of signers to hallucinate per request
     pub num_signers_per_request: usize,
+    /// Wheter to generate consecutive blocks or not
+    pub concatenate_blocks: bool,
```
Comment (matteojug) [inline signer/src/testing/storage/model.rs:528]: nit: `consecutive_blocks`? just to make it clear what's the effect

File: signer/src/request_decider.rs [hunk 12]
```diff
@@ -248,6 +268,21 @@ where
...
+            if let Err(error) = self.send_message(decision, chain_tip).await {
+                tracing::warn!(%error, "error sending deposit decision to retry, skipping");
+            }
```
Comment (matteojug) [inline signer/src/request_decider.rs:281]: nit: consider using `inspect_err` for readability

File: signer/src/testing/storage/model.rs [hunk 13]
```diff
@@ -66,8 +68,14 @@ impl TestData {
         let mut test_data = Self::new();
 
         for _ in 0..params.num_bitcoin_blocks {
-            let (next_chunk, _) = test_data.new_block(rng, signer_keys, params, None);
+            let parent = if params.concatenate_blocks {
+                test_data.bitcoin_blocks_ref.last()
+            } else {
+                None
+            };
```
Comment (matteojug) [inline signer/src/testing/storage/model.rs:75]: nit: we can use a local and avoid `bitcoin_blocks_ref`

File: signer/tests/integration/postgres.rs [hunk 14]
```diff
@@ -1381,6 +1389,106 @@ async fn fetching_deposit_request_votes() {
...
+
+    let signer_set = testing::wsts::generate_signer_set_public_keys(&mut rng, num_signers);
+
+    let mut test_data = TestData::generate(&mut rng, &signer_set, &test_model_params);
+    test_data.write_to(&pg_store).await;
+
+    let signer_pub_key = signer_set.first().unwrap();
+
+    // move the oldest deposit request to the last block
+    // so that we change its created_at time to be the latest
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1418]: nit: can you expand a bit what the resulting scenario will look like? So that is easier to follow the setup below

File: signer/tests/integration/postgres.rs [hunk 15]
```diff
@@ -1381,6 +1389,106 @@ async fn fetching_deposit_request_votes() {
...
+    test_data.deposit_requests.rotate_left(1);
+    let mut new_time_block = OffsetDateTime::now_utc() - time::Duration::minutes(15);
+    let mut new_time_request = OffsetDateTime::now_utc() - time::Duration::minutes(12);
+    // Update Bitcoin blocks
+    for (block, deposit) in test_data
```
Comment (matteojug) [inline signer/tests/integration/postgres.rs:1423]: nit: can we split this into two different for loops? It doesn't seem we are doing anything with _both_ of them, and since we also changed the coupling above with the `rotate_left` is a bit confusing.

**Review by matteojug — APPROVED** [review]
